components/server/src/auth/generic-auth-provider.ts

```diff
@@ -446,25 +449,29 @@ export class GenericAuthProvider implements AuthProvider {
 
         if (flowContext) {
             const logPayload = {
-                withIdentity: TosFlow.WithIdentity.is(flowContext) ? flowContext.candidate : undefined,
-                withUser: TosFlow.WithUser.is(flowContext) ? User.censor(flowContext.user) : undefined,
+                withIdentity: VerifyResult.WithIdentity.is(flowContext) ? flowContext.candidate : undefined,
+                withUser: VerifyResult.WithUser.is(flowContext) ? User.censor(flowContext.user) : undefined,
                 ...defaultLogPayload,
             };
-            if (
-                TosFlow.WithIdentity.is(flowContext) ||
-                (TosFlow.WithUser.is(flowContext) && flowContext.termsAcceptanceRequired)
-            ) {
-                // This is the regular path on sign up. We just went through the OAuth2 flow but didn't create a Gitpod
-                // account yet, as we require to accept the terms first.
-                log.info(context, `(${strategyName}) Redirect to /api/tos`, logPayload);
-
-                // attach the sign up info to the session, in order to proceed after acceptance of terms
-                await TosFlow.attach(request.session!, flowContext);
-
-                response.redirect(this.config.hostUrl.withApi({ pathname: "/tos", search: "mode=login" }).toString());
-                return;
+
+            if (VerifyResult.WithIdentity.is(flowContext)) {
+                log.info(context, `(${strategyName}) Creating new user and complete login.`, logPayload);
```

Comment:
– To reviewers –

This is the significant part of the whole PR. Instead of redirecting to /tos endpoint, here the two steps are inlined which would happen otherwise inside of the old /tos handler:
1. creating new account
2. finishing the login

Reply:
Previous location of these steps is https://github.com/gitpod-io/gitpod/pull/17236/files#diff-0fda9115576b7e1168858f5d4f8869e3120c3a8364cc8f43335037ebb78f6734L573

(Search for `router.get("/tos"`, if the link does not work.)